workflow/scripts/ihw-fdr-control.R

```diff
@@ -0,0 +1,116 @@
+ihw_test_grouping <- function(x){
+  tryCatch(
+    expr = {
+      ihw(pval ~ mean_obs, data = gene_data, alpha = 0.1, nbins = x)
+      return(TRUE)
+    },
+    error = function(e) {
+      print(str_c("Number of groups was set too hight, trying grouping by ", tested_number_of_groups - 1, " groups."))
+      return(FALSE)
+    })
+}
+
+while(!ihw_test_grouping(tested_number_of_groups) && tested_number_of_groups > 0){
+  tested_number_of_groups <- tested_number_of_groups -1
+  ihw_test_grouping(tested_number_of_groups)
+}
+
+if (tested_number_of_groups < number_of_groups) {
+  print(str_c("The chosen number of groups for ", level_name, " dataset was too large, instead IHW was calculated on ", tested_number_of_groups, " groups."))
+  number_of_groups <- tested_number_of_groups
+}
+
+#select the necessary data
+if(level_name == "genes-aggregated") {
+  gene_data <- gene_data %>%
+    select(-c(qval, sum_mean_obs_counts, num_aggregated_transcripts)) %>%
+    mutate(grouping = groups_by_filter(mean_obs, number_of_groups))
+} else {
+  gene_data <- gene_data %>%
+    select(ens_gene, ext_gene, pval, mean_obs) %>%
+    mutate(grouping = groups_by_filter(mean_obs, number_of_groups))
+}
+
+### diagnostic plots for covariate and grouping
+#dispersion plot
+dispersion <- ggplot(gene_data, aes(x = percent_rank(mean_obs), y = -log10(pval))) +
+  geom_point() +
+  ggtitle("IHW: scatter plot of p-values vs. mean of counts as covariate") +
+  theme(plot.title = element_text(size = 12, face = "bold", hjust = 0.5)) +
+  xlab("percent rank of mean_obs") +
+  ylab(expression(-log[10](p-value)))
+
+ggsave(snakemake@output[["dispersion"]], dispersion)
+
+#histograms
+hist_overall <- ggplot(gene_data, aes(x = pval)) +
+  geom_histogram(binwidth = 0.025, boundary = 0) +
+  xlab("p-values without grouping") +
+  ylab("density")
+
+levels(gene_data$grouping) <- paste0(rep("group ", number_of_groups), 1:number_of_groups)
+hist_groups <- ggplot(gene_data, aes(x = pval)) +
+  geom_histogram(binwidth = 0.025, boundary = 0) +
+  xlab("p-values of the individual groups") +
+  ylab("density") +
+  facet_wrap(~ grouping, nrow = number_of_groups %/% 4 + ifelse((number_of_groups %% 4 != 0), 1, 0))
+
+plots_agg_mean_obs <- ggarrange(hist_overall, hist_groups, nrow = 1)
+
+histograms <- annotate_figure(plots_agg_mean_obs,
+                              top = text_grob("IHW: histograms for p-values of mean of counts as covariate",
+                                              color = "black",
+                                              face = "bold",
+                                              size = 12))
+
+ggexport(histograms,
+         filename = snakemake@output[["histograms"]],
+         width=14)
+
+# ihw calculation
+write_tsv(gene_data, snakemake@output[["results"]])
+ihw_results_mean <- ihw(pval ~ mean_obs, data = gene_data, alpha = 0.1, nbins = tested_number_of_groups)
+ihw_mean <- as.data.frame(ihw_results_mean)
+write_tsv(ihw_mean, snakemake@output[["results"]])
```

Comment:
The second `write_tsv()` writes to the same file as the first, so it will simply overwrite its contents. Does that mean we can simply drop the first `write_tsv()` statement? Is that an artefact of script development? :D